refactor: batch slot management in decoder classes

Funatiq opened the pull request:
Highlights:
* Set up batch slots in `MakeDecodingBatchInputOutput` (outside of the decoder).
* Set up logits vector in `MakeDecodingBatchInputOutput` (outside of the decoder).
* Remove `active` vector from `decoder_batch::Input`.
* Unify the creation of decoder batch inputs in algorithm and tests.

See commit messages for details.

Funatiq committed:
refactor: batch slot management in decoder classes

- Changed `forwardBatchSlots` from a single `TensorPtr` to a `std::vector<TensorPtr>` in `decoderBuffers.h` and updated its initialization in `decoderBuffers.cpp`.
- Updated `batchSlots` in `iGptDecoderBatched.h` to a `std::vector<TensorPtr>` for better handling of batch sizes.
- Modified `mBatchSlotsDecoder` in `statefulGptDecoderBatched.h` to use a `std::vector<TensorPtr>` and adjusted its initialization in `statefulGptDecoderBatched.cpp`.
- Ensured proper reshaping of tensors in the setup methods to accommodate the new vector structure.

These changes enhance flexibility in managing tensor buffers across different batch sizes.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Enhance decoder input structure with maxDecodingEngineTokens

- Updated the `Input` class in `iGptDecoderBatched.h` to include a new parameter `maxDecodingEngineTokens` for better control over decoding limits.
- Modified the `MakeDecodingBatchInputOutput` algorithm to compute the maximum number of decoding tokens based on active slots.
- Adjusted the `GptDecoderBatched` class to utilize the new `maxDecodingEngineTokens` parameter, improving clarity in token management during decoding.
- Updated Python bindings to reflect changes in the `Input` class constructor.
- Enhanced tests to ensure proper handling of the new parameter.

These changes improve the flexibility and efficiency of the decoding process in the batch manager.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Setup batch slots outside of the decoder

- Refactored batch slot management to utilize `makeBatchSlots`, enhancing clarity and functionality in batch processing.
- Introduced `DecoderState` to `MakeDecodingBatchInputOutput` for improved state handling during decoding.
- Updated the `operator()` method to include `decoderState` as a parameter, facilitating better integration with the decoding process.
- Modified related tests to accommodate changes in batch slot handling and ensure proper functionality.

These updates improve the overall structure and efficiency of the decoding process in the batch manager.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Streamline decoder input creation and batch slot management

- Introduced a new function `createDecoderInputs` to encapsulate the logic for creating decoder inputs, improving code organization.
- Updated the `operator()` method to utilize the new `createDecoderInputs` function, simplifying the decoding input setup process.
- Removed the `maxOfActiveSlots` template function to streamline the logic for determining the maximum number of active decoding engine tokens.
- Introduced a direct calculation of `maxActiveDecodingEngineTokens` within the `createDecoderInputs` function, enhancing clarity and reducing complexity.

These changes enhance the maintainability and readability of the decoding process in the batch manager.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Update logits handling in decoder batch

- Modified the `decoder_batch::Input` to accept a vector of vectors for logits, enhancing flexibility in tensor management.
- Adjusted the `createDecoderInputs` function to accommodate the new logits structure, ensuring proper batch processing.
- Updated Python bindings to reflect changes in the `Input` class constructor, maintaining compatibility with existing interfaces.
- Refactored the `GptDecoderBatched` and `StatefulGptDecoderBatched` classes to utilize the updated logits structure, improving clarity in tensor slicing and batch size management.
- Enhanced tests to validate the new input structure and ensure correct functionality across various decoding scenarios.

These changes streamline the decoding process and improve the overall maintainability of the codebase.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: remove usage of `active` vector from prepareForward

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Removed the `active` vector from `decoder_batch::Input`

- Removed the `active` vector from the `Input` class constructor in `iGptDecoderBatched.h`, streamlining the input handling for decoding.
- Updated the `createDecoderInputs` function and related tests to reflect the changes in the `Input` class, ensuring compatibility and maintaining functionality.
- Adjusted Python bindings to accommodate the new constructor signature, enhancing clarity in the interface.

These changes improve the maintainability and readability of the decoding process in the batch manager.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: remove usage of `active` vector from gptDecoderBatchedTest

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Unify the creation of decoder batch inputs in algorithm and tests

- Added a new static method `createDecoderBatchInputs` to streamline the creation of decoder batch inputs, enhancing clarity and maintainability.
- Updated the implementation to utilize active slots directly, simplifying the logic for managing batch slots and logits.
- Refactored the `operator()` method to leverage the new input creation function, ensuring compatibility with existing decoding processes.
- Enhanced tests to validate the new input handling approach, ensuring correct functionality across various scenarios.

These changes improve the overall structure and readability of the decoding process in the batch manager.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

Funatiq commented:
/bot run

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1219](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1219/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1219](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1219/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #914](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/914/) completed with status: 'FAILURE'

Funatiq committed:
refactor: remove usage of active vector from createDecoderBatchInputs

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1223](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1223/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1223](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1223/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #917](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/917/) completed with status: 'FAILURE'

Funatiq committed:
refactor: Update maxDecoderSteps calculation

- Replaced integer division with `common::ceilDiv` for calculating `maxDecoderSteps` and `numDecoderSteps`, ensuring correct handling of token counts.

These changes enhance the robustness of the decoding batch input creation process.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Rename maxDecodingEngineTokens to maxDecoderSteps

- Updated the `Input` class in `iGptDecoderBatched.h` to rename `maxDecodingEngineTokens` to `maxDecoderSteps` for improved clarity.
- Adjusted the `createDecoderInputs` function to reflect the new naming, ensuring consistency in the decoding process.
- Modified the `GptDecoderBatched` class to utilize `maxDecoderSteps` in its logic, enhancing readability and maintainability.
- Updated Python bindings to expose the renamed parameter, maintaining compatibility with existing interfaces.

These changes enhance the clarity of the decoding parameters and improve the overall structure of the codebase.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1234](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1234/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1234](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1234/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #927](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/927/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1235](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1235/) [ run ] triggered by Bot

Funatiq commented:
/bot run --stage-list "H100_PCIe-TensorRT-1, H100_PCIe-TensorRT-3"

tensorrt-cicd commented:
[PR_Github #1237](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1237/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1237](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1237/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #930 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/930/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1490](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1490/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1490](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1490/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1111](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1111/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1502](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1502/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1502](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1502/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1119](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1119/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1585](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1585/) [ run ] triggered by Bot

dcampora commented on cpp/include/tensorrt_llm/batch_manager/decoderBuffers.h:
```diff
@@ -48,7 +48,7 @@ class DecoderInputBuffers
     TensorPtr fillValues;
     TensorPtr fillValuesDevice;
-    TensorPtr forwardBatchSlots;
+    std::vector<TensorPtr> forwardBatchSlots;
```
Why this change?

Funatiq replied:
So each tensor can be resized to the actual batch size. Previously it was a sparse 2-dim tensor. Also see [here](https://github.com/NVIDIA/TensorRT-LLM/pull/3300/files#diff-a86768fcf5826f655591d71de3615f66d0f5b55583c7329b6007cfa21c6f9e1fR73).

tensorrt-cicd commented:
[PR_Github #1585](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1585/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1187](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1187/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #1766](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1766/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1766](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1766/) [ run ] completed with state `ABORTED`

tensorrt-cicd commented:
[PR_Github #1791](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1791/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1329 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1329/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --skip-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #1906](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1906/) Bot args parsing error!

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1920](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1920/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1920](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1920/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1415](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1415/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #1952](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1952/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1952](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1952/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1437 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1437/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #2010](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2010/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #2010](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2010/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1481 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1481/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #2017](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2017/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #2017](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2017/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #1486](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1486/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #2018](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2018/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1487](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1487/) completed with status: 'SUCCESS'

Funatiq commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #2032](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2032/) [ reuse-pipeline ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #2032](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2032/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #2018](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/2018/) for commit `20e850f`
